src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -164,6 +178,71 @@ void toSerializable() {
             });
   }
 
+  @VisibleForTesting
+  void toSerializableWithExtraRead(DistributedStorage storage)
+      throws ExecutionException, CommitConflictException {
+    if (isolation != Isolation.SERIALIZABLE || strategy != SerializableStrategy.EXTRA_READ) {
+      return;
+    }
+
+    // Read set by scan is re-validated to check if there is no anti-dependency
+    Set<Key> validatedReadSetByScan = new HashSet<>();
+    for (Map.Entry<Scan, Optional<List<Key>>> entry : scanSet.entrySet()) {
+      Set<TransactionResult> currentReadSetByScan = new HashSet<>();
+      for (Result result : storage.scan(entry.getKey())) {
+        TransactionResult transactionResult = new TransactionResult(result);
+        // Ignore records that this transaction has prepared (and that are in the write set)
+        if (transactionResult.getId().equals(id)) {
+          continue;
+        }
+        currentReadSetByScan.add(transactionResult);
+      }
+
+      for (Key key : entry.getValue().get()) {
+        if (writeSet.containsKey(key)) {
+          continue;
+        }
```

Comment:
Should we also check delete-set?

Reply:
Good catch. Yes, it should be checked as well. Thank you.

Reply:
Fixed in cdb12d1.